topic-operator/scripts/topic_operator_run.sh

```diff
@@ -14,7 +14,7 @@ if [ -n "$STRIMZI_JAVA_OPTS" ]; then
     export JAVA_OPTS="${JAVA_OPTS} ${STRIMZI_JAVA_OPTS}"
 fi
 
-if [ "$STRIMZI_TLS_ENABLED" = "true" ]; then
+if [ "$STRIMZI_TLS_ENABLED" = "true" ] && [ "$STRIMZI_TLS_AUTH_ENABLED" = "true" ]; then
```

Comment:
I think the TRUSTSTORE is needed even when TLS authentication is not enabled. So it needs to be set even if `STRIMZI_TLS_ENABLED` is set to `true` but `STRIMZI_TLS_AUTH_ENABLED` is `false`.

Reply:
Sorry, I'm not a Java developer, so I can not be right here, but the current setup works perfectly for me.
Why would we need the TRUSTSTORE to be set when the authentication is disabled?
I suppose in that case java uses its default truststore (cacerts file).

Reply:
Truststore is the same as for example `caFile` in other languages. It is a set of public keys used to verify the identity of the server it is connecting to. You indeed do not need to configure it if you are fine with using the default certificates bundled with JVM (bascially if your server uses certificate signed by some trusted CA). But that doesn't make it part of TLS client authentication. It could be that this works for you but it is not the right solution and it will not work for others.

If your problem is that you do not want to set the truststore, maybe you can deal with it differently:
* Scrap the whole `TLS_AUTH_ENABLED` thing
* And change the code to something like this (not tested, just sample code):
  ```java
        if (Boolean.parseBoolean(config.get(Config.TLS_ENABLED))) {
            kafkaClientProps.setProperty(AdminClientConfig.SECURITY_PROTOCOL_CONFIG, "SSL");
            
            if (!config.get(Config.TLS_TRUSTSTORE_LOCATION).isEmpty() && !config.get(Config.TLS_TRUSTSTORE_PASSWORD).isEmpty()) {
                kafkaClientProps.setProperty(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG, config.get(Config.TLS_TRUSTSTORE_LOCATION));
                kafkaClientProps.setProperty(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG, config.get(Config.TLS_TRUSTSTORE_PASSWORD));
            }

            if (!config.get(Config.TLS_KEYSTORE_LOCATION).isEmpty() && !config.get(Config.TLS_KEYSTORE_PASSWORD).isEmpty()) {
                kafkaClientProps.setProperty(SslConfigs.SSL_KEYSTORE_LOCATION_CONFIG, config.get(Config.TLS_KEYSTORE_LOCATION));
                kafkaClientProps.setProperty(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG, config.get(Config.TLS_KEYSTORE_PASSWORD));
            }
            
            kafkaClientProps.setProperty(SslConfigs.SSL_ENDPOINT_IDENTIFICATION_ALGORITHM_CONFIG, config.get(Config.TLS_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM));
        }
  ```

This way, you do not need to change the other operator and simply if the truststore environment variables are set, the will be used. If the keystore variables are used, they will be used. (at the end, Kafka clients do not have any switch for TLS client authentication, they just do what the broker says IIRC).

Reply:
Well, I can do it, but then should we disable TRUSTSTORE and KEYSTORE generation in the topic_operator_run.sh?
I'm talking about this code block
```
if [ "$STRIMZI_TLS_ENABLED" = "true" ]; then
    if [ -z "$STRIMZI_TRUSTSTORE_LOCATION" ] && [ -z "$STRIMZI_KEYSTORE_LOCATION" ]; then
        # Generate temporary keystore password
        CERTS_STORE_PASSWORD=$(< /dev/urandom tr -dc _A-Z-a-z-0-9 | head -c32)
        export CERTS_STORE_PASSWORD

        mkdir -p /tmp/topic-operator

        # Import certificates into keystore and truststore
        "${STRIMZI_HOME}/bin/tls_prepare_certificates.sh"

        export STRIMZI_TRUSTSTORE_LOCATION=/tmp/topic-operator/replication.truststore.p12
        export STRIMZI_TRUSTSTORE_PASSWORD="$CERTS_STORE_PASSWORD"

        export STRIMZI_KEYSTORE_LOCATION=/tmp/topic-operator/replication.keystore.p12
        export STRIMZI_KEYSTORE_PASSWORD="$CERTS_STORE_PASSWORD"
    fi
fi
```

Reply:
Hm, right. You cannot disable it because it is needed. You would need to make it conditional somehow as well I guess.

Reply:
I can keep `STRIMZI_TLS_AUTH_ENABLED` variable, but change the Session.java so the `SSL_TRUSTSTORE_LOCATION_CONFIG` and `SSL_TRUSTSTORE_PASSWORD_CONFIG` will be set if it's set as `STRIMZI_TRUSTSTORE_LOCATION` and `STRIMZI_TRUSTSTORE_PASSWORD` env vars.

Like this
```
if (Boolean.parseBoolean(config.get(Config.TLS_ENABLED))) {
            kafkaClientProps.setProperty(AdminClientConfig.SECURITY_PROTOCOL_CONFIG, "SSL");
            kafkaClientProps.setProperty(SslConfigs.SSL_ENDPOINT_IDENTIFICATION_ALGORITHM_CONFIG, config.get(Config.TLS_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM));

            if (!config.get(Config.TLS_TRUSTSTORE_LOCATION).isEmpty() && !config.get(Config.TLS_TRUSTSTORE_PASSWORD).isEmpty()) {
                kafkaClientProps.setProperty(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG, config.get(Config.TLS_TRUSTSTORE_LOCATION));
                kafkaClientProps.setProperty(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG, config.get(Config.TLS_TRUSTSTORE_PASSWORD));
            }
            if (Boolean.parseBoolean(config.get(Config.TLS_AUTH_ENABLED))) {
                kafkaClientProps.setProperty(SslConfigs.SSL_KEYSTORE_LOCATION_CONFIG, config.get(Config.TLS_KEYSTORE_LOCATION));
                kafkaClientProps.setProperty(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG, config.get(Config.TLS_KEYSTORE_PASSWORD));
            }
        }
```


And the `topic-operator-run.sh` script will remain the same

```
if [ "$STRIMZI_TLS_ENABLED" = "true" ] && [ "$STRIMZI_TLS_AUTH_ENABLED" = "true" ]; then
    if [ -z "$STRIMZI_TRUSTSTORE_LOCATION" ] && [ -z "$STRIMZI_KEYSTORE_LOCATION" ]; then
        # Generate temporary keystore password
        CERTS_STORE_PASSWORD=$(< /dev/urandom tr -dc _A-Z-a-z-0-9 | head -c32)
        export CERTS_STORE_PASSWORD

        mkdir -p /tmp/topic-operator

        # Import certificates into keystore and truststore
        "${STRIMZI_HOME}/bin/tls_prepare_certificates.sh"

        export STRIMZI_TRUSTSTORE_LOCATION=/tmp/topic-operator/replication.truststore.p12
        export STRIMZI_TRUSTSTORE_PASSWORD="$CERTS_STORE_PASSWORD"

        export STRIMZI_KEYSTORE_LOCATION=/tmp/topic-operator/replication.keystore.p12
        export STRIMZI_KEYSTORE_PASSWORD="$CERTS_STORE_PASSWORD"
    fi
fi
```

This way both full strimzi and standalone deployments will work, right?

Reply:
I think this possibly breaks it for other users who want to have just the truststore configured.

Reply:
Why? If they have specified the `STRIMZI_TRUSTSTORE_LOCATION` and `STRIMZI_TRUSTSTORE_LOCATION` - then it'll be set as a property according to my latest example of Session.java:
```
if (Boolean.parseBoolean(config.get(Config.TLS_ENABLED))) {
            kafkaClientProps.setProperty(AdminClientConfig.SECURITY_PROTOCOL_CONFIG, "SSL");
            kafkaClientProps.setProperty(SslConfigs.SSL_ENDPOINT_IDENTIFICATION_ALGORITHM_CONFIG, config.get(Config.TLS_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM));

            if (!config.get(Config.TLS_TRUSTSTORE_LOCATION).isEmpty() && !config.get(Config.TLS_TRUSTSTORE_PASSWORD).isEmpty()) {
                kafkaClientProps.setProperty(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG, config.get(Config.TLS_TRUSTSTORE_LOCATION));
                kafkaClientProps.setProperty(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG, config.get(Config.TLS_TRUSTSTORE_PASSWORD));
            }
            if (Boolean.parseBoolean(config.get(Config.TLS_AUTH_ENABLED))) {
                kafkaClientProps.setProperty(SslConfigs.SSL_KEYSTORE_LOCATION_CONFIG, config.get(Config.TLS_KEYSTORE_LOCATION));
                kafkaClientProps.setProperty(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG, config.get(Config.TLS_KEYSTORE_PASSWORD));
            }
        }
```

Reply:
We can even specify a conditional set of keystore properties even if `TLS_AUTH_ENABLED` is set to `true` (it's set to `true` by default)

Reply:
I'm sorry, but the truststore is simply not part of TLS authentication and should not be gated by it. I appreciate that you are trying to contribute, but it needs to be done right. There might be users using only TLS without authentication but who still need to provide the truststore. So it needs to be possible without enabling the TLS client authentication.

Reply:
This is actually what I'm talking about.
Regarding my latest example they could specify `TRUSTORE` location and password even if the `TLS_AUTH_ENABLED` is set to `false`.

Maybe I got you wrong, so could you please clarify if this is the case.

P.S. Thanks for your active participation <3

Reply:
The last Java code looks fine (although I guess we do not need `TLS_AUTH_ENABLED` really here since we can check the keystore the same way as truststore?). But the Bash code has it still gated both through `TLS_AUTH_ENABLED`

Reply:
Well, as far as I see, the bash code generates certificates if `STRIMZI_TRUSTSTORE_LOCATION` and `STRIMZI_KEYSTORE_LOCATION` are empty. This certificate generation only works with the full strimzi deployment. It doesn't work for standalone topic operator deployment.
The idea is:
If you use the full strimzi deployment (with kafka, zookeper etc. managed by strimzi in k8s) and `TLS_ENABLED` set to `true` then you'll always have both TLS encryption and authentication enabled. And the bash script `if` condition will pass. They couldn't disable authentication (why would they want to do it?)
If you use the standalone topic operator deployment - then the bash conditions will never pass because you have either `TLS_AUTH_ENABLED` set to `true` with `TRUSTSTORE` and `KEYSTORE` specified or `TLS_AUTH_ENABLED` set to `false` with `TRUSTSTORE` either specified or not.

Do you want me to also implement TLS_AUTH switch for the full strimzi deployment?

Reply:
No, you can have standalone users who:
* Want to use TLS encryption
* Want to use it with self signed certificate (i.e. they need the truststore to be generated)
* But do not want to use TLS authentication => so `TLS_AUTH_ENABLED=false` and no keystore